Fixes #5 - Add/remove project to/from dev server

akerekes opened the pull request:
This is the code copied over from GPE and modified to compile and work with the Cloud SDK's changed CLI.
There are still several features missing/not working as expected, I'll file issues about the ones I am aware of.
//TODOs are present in the code, I'll replace them with issues as well.

@elharo @briandealwis PTAL

elharo commented on plugins/com.google.cloud.tools.eclipse.appengine.localserver.test/src/com/google/cloud/tools/eclipse/appengine/localserver/CloudSdkUtilsTest.java:
```diff
@@ -0,0 +1,116 @@
+ */
+@RunWith(MockitoJUnitRunner.class)
+public class CloudSdkUtilsTest {
+  @Mock 
```
No firm rule, but as long as the lines aren't too long consider putting @Mock on same line

elharo commented on plugins/com.google.cloud.tools.eclipse.appengine.localserver.test/src/com/google/cloud/tools/eclipse/appengine/localserver/CloudSdkUtilsTest.java:
```diff
@@ -0,0 +1,116 @@
+   */
+  @Test
+  public void testHasGcloudMavenPlugin_noGcloudPlugin() {
+    List<Plugin> plugins = createPluginList("myArtifcatId1", "myGroupId1", "myArtifcatId2",
```
spelling nit: myArtifactId1, myArtifactId2

elharo commented on plugins/com.google.cloud.tools.eclipse.appengine.localserver.test/src/com/google/cloud/tools/eclipse/appengine/localserver/CloudSdkUtilsTest.java:
```diff
@@ -0,0 +1,116 @@
+  // that is mocked here
+  @SuppressWarnings("restriction")
+  @Test
+  public void testhasCloudSdkFacet_nonFacetProject() throws CoreException {
```
nit: testHas

elharo commented on plugins/com.google.cloud.tools.eclipse.appengine.localserver.test/src/com/google/cloud/tools/eclipse/appengine/localserver/GCloudCommandDelegateTest.java:
```diff
@@ -0,0 +1,141 @@
+/*******************************************************************************
+ * Copyright 2015 Google Inc. All Rights Reserved.
```
2016

elharo commented on plugins/com.google.cloud.tools.eclipse.appengine.localserver.test/src/com/google/cloud/tools/eclipse/appengine/localserver/CloudSdkUtilsTest.java:
```diff
@@ -0,0 +1,116 @@
+/*******************************************************************************
+ * Copyright 2015 Google Inc. All Rights Reserved.
```
2016

elharo commented on plugins/com.google.cloud.tools.eclipse.appengine.localserver.test/src/com/google/cloud/tools/eclipse/appengine/localserver/GCloudCommandDelegateTest.java:
```diff
@@ -0,0 +1,141 @@
+  
+  @Test
+  public void testIsComponent_updateAvailable() {
+    String output =  createOutput("Update Available");
```
nit: one space after =

elharo commented on plugins/com.google.cloud.tools.eclipse.appengine.localserver.test/src/com/google/cloud/tools/eclipse/appengine/localserver/GCloudCommandDelegateTest.java:
```diff
@@ -0,0 +1,141 @@
+  
+  @Test
+  public void testIsComponentInstalled_notInstalled() {
+    String output =  createOutput("Not Installed");
```
nit: one space after =

elharo commented on plugins/com.google.cloud.tools.eclipse.appengine.localserver.test/src/com/google/cloud/tools/eclipse/appengine/localserver/GCloudCommandDelegateTest.java:
```diff
@@ -0,0 +1,141 @@
+  public void testIsComponent_updateAvailable() {
+    String output =  createOutput("Update Available");
+    assertTrue(GCloudCommandDelegate.isComponentInstalled(output,
+        GCloudCommandDelegate.APP_ENGINE_COMPONENT_NAME));
```
four space indent for continuations

elharo commented on plugins/com.google.cloud.tools.eclipse.appengine.localserver.test/src/com/google/cloud/tools/eclipse/appengine/localserver/GCloudCommandDelegateTest.java:
```diff
@@ -0,0 +1,141 @@
+  public void testIsComponent_invalidComponent() {
+    String output =  createOutput("Installed");
+    assertFalse(GCloudCommandDelegate.isComponentInstalled(output,
+        "invalid-component-name"));
```
four space indent for continuations

elharo commented on plugins/com.google.cloud.tools.eclipse.appengine.localserver/plugin.xml:
```diff
@@ -1,5 +1,5 @@
 <?xml version="1.0" encoding="UTF-8"?>
-<?eclipse version="3.4"?>
+<?eclipse version="3.2"?>
```
3.2?

akerekes replied:
@elharo That is the dtd version, it has not much correlation with the Eclipse versions, for Mars it is 3.2:
http://help.eclipse.org/mars/index.jsp?topic=%2Forg.eclipse.platform.doc.isv%2Freference%2Fmisc%2Fplugin_manifest.html

briandealwis replied:
Um, I think the docs aren't up to date.  Creating a new plug-in with PDE has 3.4:

```
<?eclipse version="3.4"?>
```

elharo commented on plugins/com.google.cloud.tools.eclipse.appengine.localserver/plugin.xml:
```diff
@@ -100,4 +100,207 @@
+                  <requires facet="jst.web" version="3.0"/>
+                  <requires facet="jst.web" version="3.1"/>
+              </or>
+              <requires facet="java" version="[1.6"/>
```
1.7?

akerekes replied:
@elharo the syntax is correct:
https://eclipse.org/articles/Article-BuildingProjectFacets/tutorial.html#version.expressions
I've changed it to [1.7

briandealwis replied:
I think it should be "1.7]" — anything up to and including 1.7

elharo commented on plugins/com.google.cloud.tools.eclipse.appengine.localserver/plugin.xml:
```diff
@@ -100,4 +100,207 @@
+          <facet id="jst.web" version="2.2,2.3,2.4,2.5,3.1"/>
+          <facet id="jst.utility"  version="1.0"/>
+          <facet id="jst.webfragment" version="3.0" />
+          <facet id="java" version="[1.6" />
```
1.7?

briandealwis replied:
"1.7]"

elharo commented on plugins/com.google.cloud.tools.eclipse.appengine.localserver/src/com/google/cloud/tools/eclipse/appengine/localserver/Activator.java:
```diff
@@ -0,0 +1,117 @@
+   * message and title. Ensures that the error dialog is opened on the UI
+   * thread.
+   *
+   * @param parent
```
these could all go to one line

elharo commented on plugins/com.google.cloud.tools.eclipse.appengine.localserver/src/com/google/cloud/tools/eclipse/appengine/localserver/CloudSdkUtils.java:
```diff
@@ -0,0 +1,124 @@
+/*******************************************************************************
+ * Copyright 2015 Google Inc. All Rights Reserved.
```
2016

elharo commented on plugins/com.google.cloud.tools.eclipse.appengine.localserver/src/com/google/cloud/tools/eclipse/appengine/localserver/CloudSdkUtils.java:
```diff
@@ -0,0 +1,124 @@
+ */
+@SuppressWarnings("restriction") // For FacetUtil
+public class CloudSdkUtils {
+  public static final String APPENGINE_GROUP_ID = "com.google.appengine";
```
should this be updated to com.google.cloud.tools.eclipse?

akerekes replied:
@elharo This code checks for the presence of the gcloud-maven-plugin, whose group ID is `com.google.appengine`.
On the other hand we'll need to extend this functionality to support the new Maven plugin if needed once it's ready to be used.

elharo commented on plugins/com.google.cloud.tools.eclipse.appengine.localserver/src/com/google/cloud/tools/eclipse/appengine/localserver/GCloudCommandDelegate.java:
```diff
@@ -0,0 +1,178 @@
+/*******************************************************************************
+ * Copyright 2015 Google Inc. All Rights Reserved.
```
2016

elharo commented on plugins/com.google.cloud.tools.eclipse.appengine.localserver/src/com/google/cloud/tools/eclipse/appengine/localserver/deploy/CloudSdkDeployProjectHandler.java:
```diff
@@ -0,0 +1,158 @@
+/*******************************************************************************
+ * Copyright 2015 Google Inc. All Rights Reserved.
```
2016

elharo commented on plugins/com.google.cloud.tools.eclipse.appengine.localserver/src/com/google/cloud/tools/eclipse/appengine/localserver/deploy/CloudSdkDeployProjectHandler.java:
```diff
@@ -0,0 +1,158 @@
+    boolean hasLoggedInUsers;
+    try {
+      hasLoggedInUsers = GCloudCommandDelegate.hasLoggedInUsers(project, runtime);
+    } catch (Exception e) {
```
be more specific about exceptions

elharo commented on plugins/com.google.cloud.tools.eclipse.appengine.localserver/src/com/google/cloud/tools/eclipse/appengine/localserver/deploy/CloudSdkDeployProjectHandler.java:
```diff
@@ -0,0 +1,158 @@
+          if (exitCode != 0) {
+            Activator.logAndDisplayError(null, TITLE, "Cloud SDK deploy action terminated with exit code " + exitCode);
+          }
+        } catch (Throwable e) {
```
don't catch throwable, per effective java

elharo commented on plugins/com.google.cloud.tools.eclipse.appengine.localserver/src/com/google/cloud/tools/eclipse/appengine/localserver/deploy/CloudSdkProjectPropertyTester.java:
```diff
@@ -0,0 +1,41 @@
+/*******************************************************************************
+ * Copyright 2015 Google Inc. All Rights Reserved.
```
2016

elharo commented on plugins/com.google.cloud.tools.eclipse.appengine.localserver/src/com/google/cloud/tools/eclipse/appengine/localserver/facet/CloudSdkFacetInstallDelegate.java:
```diff
@@ -0,0 +1,47 @@
+/*******************************************************************************
+ * Copyright 2015 Google Inc. All Rights Reserved.
```
2016

elharo commented on plugins/com.google.cloud.tools.eclipse.appengine.localserver/src/com/google/cloud/tools/eclipse/appengine/localserver/facet/CloudSdkProjectConfigurator.java:
```diff
@@ -0,0 +1,81 @@
+   * Cloud SDK runtimes to the list of targeted runtimes for
+   * {@code facetedProject}.
+   *
+   * @param facetedProject
```
one line for javadoc @param

elharo commented on plugins/com.google.cloud.tools.eclipse.appengine.localserver/src/com/google/cloud/tools/eclipse/appengine/localserver/runtime/CloudSdkRuntime.java:
```diff
@@ -0,0 +1,54 @@
+        return new Status(IStatus.ERROR, Activator.PLUGIN_ID, 0, "The Cloud SDK/App Engine component is not installed",
+            null);
+      }
+    } catch (Exception e) {
```
more specific exception

elharo commented on plugins/com.google.cloud.tools.eclipse.appengine.localserver/src/com/google/cloud/tools/eclipse/appengine/localserver/server/CloudSdkLaunchConfigurationDelegate.java:
```diff
@@ -0,0 +1,213 @@
+        cloudSdkServer.getApiHost(), cloudSdkServer.getApiPort(), debugPort);
+    Activator.logInfo("Executing: " + commands);
+
+    Process p = null;
```
try not to split variables across try blocks. In general more code should go into a try, not less

elharo commented on plugins/com.google.cloud.tools.eclipse.appengine.localserver/src/com/google/cloud/tools/eclipse/appengine/localserver/server/CloudSdkServer.java:
```diff
@@ -0,0 +1,213 @@
+/*******************************************************************************
+ * Copyright 2015 Google Inc. All Rights Reserved.
```
2016

elharo commented on plugins/com.google.cloud.tools.eclipse.appengine.localserver/src/com/google/cloud/tools/eclipse/appengine/localserver/server/CloudSdkServer.java:
```diff
@@ -0,0 +1,213 @@
+      }
+    }
+
+    assert(modules.size() >= remove.length);
```
no asserts please

elharo commented on plugins/com.google.cloud.tools.eclipse.appengine.localserver/src/com/google/cloud/tools/eclipse/appengine/localserver/server/CloudSdkServerBehaviour.java:
```diff
@@ -0,0 +1,224 @@
+      String url = "http://" + adminHost + ":" + adminPort;
+      pingThread = new PingThread(getServer(), url, -1, this);
+      pingThread.start();
+    } catch (Throwable e) {
```
don't catch Throwable

elharo commented on plugins/com.google.cloud.tools.eclipse.appengine.localserver/src/com/google/cloud/tools/eclipse/appengine/localserver/server/CloudSdkServerBehaviour.java:
```diff
@@ -0,0 +1,224 @@
+      }
+
+      if (pingThread != null) {
+        pingThread.stop();
```
probably OK here; but in general stopping a thread is strongly discouraged for safety reasons

elharo commented on plugins/com.google.cloud.tools.eclipse.appengine.localserver/src/com/google/cloud/tools/eclipse/appengine/localserver/server/CloudSdkServerBehaviour.java:
```diff
@@ -0,0 +1,224 @@
+      }
+
+      setServerState(IServer.STATE_STOPPED);
+    } catch (Throwable e) {
```
don't catch Throwable

elharo commented on plugins/com.google.cloud.tools.eclipse.appengine.localserver/src/com/google/cloud/tools/eclipse/appengine/localserver/server/CloudSdkServerBehaviour.java:
```diff
@@ -0,0 +1,224 @@
+  }
+
+  private boolean isPortAvailable(int port) {
+    Socket socket = null;
```
try with resources if we're using Java 7

elharo commented on plugins/com.google.cloud.tools.eclipse.appengine.localserver/src/com/google/cloud/tools/eclipse/appengine/localserver/server/CloudSdkServerBehaviour.java:
```diff
@@ -0,0 +1,224 @@
+
+      URL url = new URL("http", ad, port, "/quit");
+      connection = (HttpURLConnection) url.openConnection();
+      connection.setDoOutput(true);
```
I don't think you want this; delete this line

elharo commented on plugins/com.google.cloud.tools.eclipse.appengine.localserver/src/com/google/cloud/tools/eclipse/appengine/localserver/server/PingThread.java:
```diff
@@ -0,0 +1,126 @@
+   * @param behaviour
+   *          the {@link ServerBehaviourDelegate} of {@code server}
+   */
+  public PingThread(IServer server, String url, int maxPings, CloudSdkServerBehaviour behaviour) {
```
Ick. PingThread is not a Thread at all. Please rename.

elharo commented on plugins/com.google.cloud.tools.eclipse.appengine.localserver/src/com/google/cloud/tools/eclipse/appengine/localserver/server/PingThread.java:
```diff
@@ -0,0 +1,126 @@
+  private static final int PING_INTERVAL = 250;
+
+  // maximum number of pings before giving up
+  private int maxPings;
```
make these final if at all possible

elharo commented on plugins/com.google.cloud.tools.eclipse.appengine.localserver/src/com/google/cloud/tools/eclipse/appengine/localserver/server/PingThread.java:
```diff
@@ -0,0 +1,126 @@
+    }
+    while (!stop) {
+      try {
+        if (count == maxPings) {
```
>=

elharo commented on plugins/com.google.cloud.tools.eclipse.appengine.localserver/src/com/google/cloud/tools/eclipse/appengine/localserver/server/PingThread.java:
```diff
@@ -0,0 +1,126 @@
+    int count = 0;
+    try {
+      Thread.sleep(PING_DELAY);
+    } catch (Throwable e) {
```
InterruptedException ex

elharo commented on plugins/com.google.cloud.tools.eclipse.appengine.localserver/src/com/google/cloud/tools/eclipse/appengine/localserver/server/PingThread.java:
```diff
@@ -0,0 +1,126 @@
+          behaviour.setServerStarted();
+        }
+        stop = true;
+      } catch (Throwable e) {
```
don't catch throwable

elharo commented on plugins/com.google.cloud.tools.eclipse.appengine.localserver/src/com/google/cloud/tools/eclipse/appengine/localserver/server/ServerFlagsInfo.java:
```diff
@@ -0,0 +1,50 @@
+   * "flags" in the server-flags.json.
+   */
+  public static class Flag {
+    private String name;
```
I don't see the setters for these fields. What am I missing?

briandealwis replied:
@elharo Filled in by com.google.gson when deserializing .json files

elharo commented on plugins/com.google.cloud.tools.eclipse.appengine.localserver/src/com/google/cloud/tools/eclipse/appengine/localserver/ui/CloudSdkRuntimeWizardFragment.java:
```diff
@@ -0,0 +1,201 @@
+    wizard = handle;
+    runtime = getRuntimeDelegate();
+
+    String title = getRuntimeTitle();
```
make this class final if possible

elharo commented on plugins/com.google.cloud.tools.eclipse.appengine.localserver/src/com/google/cloud/tools/eclipse/appengine/localserver/ui/CloudSdkServerLaunchTab.java:
```diff
@@ -0,0 +1,158 @@
+// pull the appropriate
+// program args from Maven into the launch config, or vice-versa..
+public class CloudSdkServerLaunchTab extends AbstractLaunchConfigurationTab {
+  protected Text prgmFlagText;
```
programFlagText

elharo commented on plugins/com.google.cloud.tools.eclipse.appengine.localserver/src/com/google/cloud/tools/eclipse/appengine/localserver/ui/ServerFlagSelectionDialog.java:
```diff
@@ -0,0 +1,144 @@
+    Object[] selected = getResult();
+    if (selected != null && selected.length == 1) {
+      Flag variable = (Flag) selected[0];
+      StringBuffer buffer = new StringBuffer();
```
StringBuilder

elharo commented on plugins/com.google.cloud.tools.eclipse.appengine.localserver/src/com/google/cloud/tools/eclipse/util/MavenUtils.java:
```diff
@@ -0,0 +1,37 @@
+
+  /**
+   * Returns <code>true</code> if the given project has the Maven 2 nature. This
+   * checks for the old maven nature (till m2Eclipse 0.12) and the new Maven
```
version numbers here seem off

akerekes replied:
@elharo Not sure what you mean. On the other hand maybe we can get rid of checking for the old plugin as it has been released 6 years ago...

elharo replied:
m2eclipse is now at 1.6.2.

akerekes replied:
0.12 was the last version of the plugin from Sonatype. Since then it is released by Eclipse and uses a new ID for the nature. It is at version 1.6.2, which is also included in the `m2Eclipse 1.0.0 and up` phrase so I think we're good here.

elharo commented on plugins/com.google.cloud.tools.eclipse.appengine.localserver/src/com/google/cloud/tools/eclipse/util/OSUtilities.java:
```diff
@@ -0,0 +1,43 @@
+public class OSUtilities {
+
+  public static boolean isMac() {
+    String os = System.getProperty("os.name").toLowerCase();
```
specify locale for toLowerCase

elharo replied:
ping

elharo commented on plugins/com.google.cloud.tools.eclipse.appengine.localserver/src/com/google/cloud/tools/eclipse/util/OSUtilities.java:
```diff
@@ -0,0 +1,43 @@
+  }
+
+  public static boolean isUnix() {
+    String os = System.getProperty("os.name").toLowerCase();
```
specify locale for toLowerCase

elharo replied:
ping

elharo commented on plugins/com.google.cloud.tools.eclipse.appengine.localserver/src/com/google/cloud/tools/eclipse/util/OSUtilities.java:
```diff
@@ -0,0 +1,43 @@
+  }
+
+  public static boolean isWindows() {
+    String os = System.getProperty("os.name").toLowerCase();
```
specify locale for toLowerCase

elharo replied:
ping

elharo commented on plugins/com.google.cloud.tools.eclipse.appengine.localserver/src/com/google/cloud/tools/eclipse/util/ProcessUtilities.java:
```diff
@@ -0,0 +1,256 @@
+   *          optional, will be given the process right after it is started
+   * @return process exit code
+   * 
+   * @throws IOException
```
no need to include @throws unless you have something to say here

elharo commented on plugins/com.google.cloud.tools.eclipse.appengine.localserver/src/com/google/cloud/tools/eclipse/util/ProcessUtilities.java:
```diff
@@ -0,0 +1,256 @@
+          newPathEnvVar.append(path);
+        }
+      }
+      // finally, set the evn variable on the process builder
```
evn --> environment

elharo commented on plugins/com.google.cloud.tools.eclipse.appengine.localserver/src/com/google/cloud/tools/eclipse/util/ProcessUtilities.java:
```diff
@@ -0,0 +1,256 @@
+
+        public void run() {
+          BufferedReader processReader = new BufferedReader(new InputStreamReader(process.getInputStream()));
+          PrintWriter outputWriter = new PrintWriter(outputStream, true);
```
PrintWriter is risky; swallows exceptions

elharo commented on plugins/com.google.cloud.tools.eclipse.appengine.localserver/src/com/google/cloud/tools/eclipse/util/TerminateAction.java:
```diff
@@ -0,0 +1,62 @@
+ *******************************************************************************/
+package com.google.cloud.tools.eclipse.util;
+
+
```
delete extra blank line

briandealwis commented on plugins/com.google.cloud.tools.eclipse.appengine.localserver/src/com/google/cloud/tools/eclipse/appengine/localserver/facet/CloudSdkFacetInstallDelegate.java:
```diff
@@ -0,0 +1,47 @@
+      throws CoreException {
+    if (!CloudSdkUtils.hasGcloudMavenPlugin(project)) {
+      // This exception will prevent the facet from being added to the project
+      throw new CoreException(new Status(IStatus.ERROR, Activator.PLUGIN_ID, "Cannot install Cloud SDK facet.\n"
```
So I thought we were going to allow native deployments?

akerekes replied:
@briandealwis you're right, I file an issue, but keep this code as it is to not hinder the merge too long.

briandealwis commented on plugins/com.google.cloud.tools.eclipse.appengine.localserver/src/com/google/cloud/tools/eclipse/appengine/localserver/deploy/CloudSdkDeployProjectHandler.java:
```diff
@@ -0,0 +1,158 @@
+    }
+
+    final IPath[] fileSystemPath = new IPath[1];
+    Display.getDefault().syncExec(new Runnable() {
```
Never use `Display.getDefault()` — if there is no display, this creates a display object, which is not the right thing for 99.999% of the time. If unavailable, use PlatformUI.getWorkbench().getDisplay().

briandealwis commented on plugins/com.google.cloud.tools.eclipse.appengine.localserver/src/com/google/cloud/tools/eclipse/appengine/localserver/deploy/CloudSdkDeployProjectHandler.java:
```diff
@@ -0,0 +1,158 @@
+    final IPath[] fileSystemPath = new IPath[1];
+    Display.getDefault().syncExec(new Runnable() {
+      public void run() {
+        Shell shell = PlatformUI.getWorkbench().getActiveWorkbenchWindow().getShell();
```
It's possible for `getActiveWorkbenchWindow()` to return null.  Here's it's better to use `PlatformUI.getWorkbench().getDisplay().getActiveShell()`.

briandealwis commented on plugins/com.google.cloud.tools.eclipse.appengine.localserver/src/com/google/cloud/tools/eclipse/appengine/localserver/deploy/CloudSdkProjectPropertyTester.java:
```diff
@@ -0,0 +1,41 @@
+ * enable the "Cloud SDK Deploy Project" context menu item for an Eclipse
+ * project.
+ */
+public class CloudSdkProjectPropertyTester extends PropertyTester {
```
If we're just looking for a project facet, then we can remove this and just use the WTP projectFacet tester.

briandealwis commented on plugins/com.google.cloud.tools.eclipse.appengine.localserver/src/com/google/cloud/tools/eclipse/appengine/localserver/server/CloudSdkLaunchConfigurationDelegate.java:
```diff
@@ -0,0 +1,213 @@
+      return;
+    }
+
+    final IModule[] modules = server.createWorkingCopy().getModules();
```
Does this really need to create a working copy?

briandealwis commented on eclipse/mars/pom.xml:
```diff
@@ -12,4 +12,12 @@
+  <dependencies>
+    <dependency>
+      <groupId>org.easymock</groupId>
+      <artifactId>easymock</artifactId>
```
I'm rather surprised that this actually does anything

briandealwis commented on plugins/com.google.cloud.tools.eclipse.appengine.localserver/META-INF/MANIFEST.MF:
```diff
@@ -16,4 +19,16 @@ Require-Bundle: org.eclipse.wst.server.core;bundle-version="1.7.0",
  org.eclipse.jst.common.project.facet.core;bundle-version="1.4.400",
  org.eclipse.jst.common.project.facet.ui;bundle-version="1.4.510",
  javax.servlet;bundle-version="3.1.0";visibility:=reexport,
```
Re-exporting is generally not a good idea.  Why do you think this is necessary here?

akerekes replied:
Removed.

briandealwis commented on plugins/com.google.cloud.tools.eclipse.appengine.localserver/plugin.properties:
```diff
@@ -0,0 +1,14 @@
+cloudsdkfacetlabel=Google Cloud SDK
+cloudsdkfacetdescription=Adds support for App Engine through the Google Cloud SDK
+
+pluginName=Google Cloud SDK WTP Tools
```
This description seems more general than it should.  _Local development support for AppEngine-Standard for Eclipse WTP_?

briandealwis commented on plugins/com.google.cloud.tools.eclipse.appengine.localserver/plugin.xml:
```diff
@@ -100,4 +100,207 @@
+      <runtimeType
+            class="com.google.cloud.tools.eclipse.appengine.localserver.runtime.CloudSdkRuntime"
+            description="%runtimeTypeDescription"
+            id="com.google.cloud.tools.eclipse.runtime.id1"
```
This `id` doesn't convey what I thought it might.

akerekes replied:
#79

briandealwis commented on plugins/com.google.cloud.tools.eclipse.appengine.localserver/plugin.xml:
```diff
@@ -100,4 +100,207 @@
+            name="%runtimeTypeName"
+            vendor="%runtimeTypeVendor"
+            version="1">
+            <moduleType types="jst.web" versions="2.2, 2.3, 2.4, 2.5, 3.1"/>
```
GAE-Standard doesn't supports 3.x? https://code.google.com/p/googleappengine/issues/detail?id=3091

akerekes replied:
@briandealwis It does not support the full 3.1 spec, but it supports annotations, so my thinking here was to allow for 3.1 so Eclipse does not warn about annotations. But I agree, that users might be tempted to use other parts of the spec that is not supported by AE Standard Environment to find it out only later, and possibly not in a very straightforward way.

I'd keep the code as is, but file an issue to find a solution later.

briandealwis commented on plugins/com.google.cloud.tools.eclipse.appengine.localserver/plugin.xml:
```diff
@@ -100,4 +100,207 @@
+      </fragment>
+   </extension>
+   <extension point="org.eclipse.wst.common.project.facet.core.facets">
+      <project-facet id="com.google.cloud.tools.eclipse.facet">
```
This facet id doesn't convey much to me either

akerekes replied:
https://github.com/GoogleCloudPlatform/gcloud-eclipse-tools/issues/79

elharo replied:
I'm not sure what you mean by thumbs up. That's usually an LGTM from a reviewer.

briandealwis commented on plugins/com.google.cloud.tools.eclipse.appengine.localserver/plugin.xml:
```diff
@@ -100,4 +100,207 @@
+                  <requires facet="jst.web" version="2.3"/>
+                  <requires facet="jst.web" version="2.4"/>
+                  <requires facet="jst.web" version="2.5"/>
+                  <requires facet="jst.web" version="3.0"/>
```
And I thought AE Standard didn't support Servlet 3.x

akerekes replied:
https://github.com/GoogleCloudPlatform/gcloud-eclipse-tools/issues/79

briandealwis commented on plugins/com.google.cloud.tools.eclipse.appengine.localserver/plugin.xml:
```diff
@@ -100,4 +100,207 @@
+      <supported>
+          <runtime-component id="com.google.cloud.tools.eclipse.runtime.id" version="1"/>
+          <facet id="com.google.cloud.tools.eclipse.facet" version="1"/>
+          <facet id="jst.web" version="2.2,2.3,2.4,2.5,3.1"/>
```
Not 3.1

briandealwis commented on plugins/com.google.cloud.tools.eclipse.appengine.localserver/plugin.xml:
```diff
@@ -100,4 +100,207 @@
+   <extension
+         point="org.eclipse.core.expressions.propertyTesters">
+      <propertyTester
+            class="com.google.cloud.tools.eclipse.appengine.localserver.deploy.CloudSdkProjectPropertyTester"
```
A few remarks:
- rather than define our own property tester, it's probably better from an extensibility perspective to provide an expression _definition_ (`org.eclipse.core.expressions.definition`) so we can vary the implementation
- WTP provides a [facet property tester](https://eclipse.org/articles/Article-BuildingProjectFacets/tutorial.html#property.tester)
- I think some of this is misnamed. CloudSDKProject seems to coarse grained, and I think we'll need to be able to distinguish between AE-Standard and AE-Flexible

akerekes replied:
issue to use expression definition or facet property tester: #81 
better naming: #79

briandealwis commented on plugins/com.google.cloud.tools.eclipse.appengine.localserver/plugin.properties:
```diff
@@ -0,0 +1,14 @@
+pluginName=Google Cloud SDK WTP Tools
+providerName=Google, Inc.
+
+runtimeTypeDescription=Cloud SDK Runtime
```
I'd have expected these runtime type strings to mention AppEngine

akerekes replied:
https://github.com/GoogleCloudPlatform/gcloud-eclipse-tools/issues/79

briandealwis commented:
I haven't finished going through the changes, but here are some higher-level questions:
- is 'Cloud SDK' really the right name to present?  This is adding support for AppEngine-Std (AE1), but I'd expect there are going to be changes needed for AE-Flex.  There is a lot falling under GCloud SDK, like GCE, GKE, BigQuery, etc.
- the server-type identifiers don't need 'eclipse' as in the facet `com.google.cloud.tools.eclipse.facet` or the server runtime id `com.google.cloud.tools.eclipse.runtime.id1`

elharo commented on plugins/com.google.cloud.tools.eclipse.appengine.localserver/src/com/google/cloud/tools/eclipse/appengine/localserver/GCloudCommandDelegate.java:
```diff
@@ -0,0 +1,178 @@
+   *          the location of the Cloud SDK
+   * @return true if the Cloud SDK and App Engine component have been installed
+   *         and false otherwise
+   * @throws IOException
```
no @throws without description

elharo commented on plugins/com.google.cloud.tools.eclipse.appengine.localserver/src/com/google/cloud/tools/eclipse/appengine/localserver/GCloudCommandDelegate.java:
```diff
@@ -0,0 +1,178 @@
+   * Returns true if the Cloud SDK and the App Engine component have been
+   * installed. Returns false otherwise.
+   *
+   * @param sdkLocation
```
pull in to one line

elharo commented on plugins/com.google.cloud.tools.eclipse.appengine.localserver/src/com/google/cloud/tools/eclipse/appengine/localserver/GCloudCommandDelegate.java:
```diff
@@ -0,0 +1,178 @@
+   *          an Eclipse project
+   * @param serverRuntime
+   *          a Cloud SDK runtime
+   * @return
```
what?

patflynn commented on plugins/com.google.cloud.tools.eclipse.appengine.localserver/src/com/google/cloud/tools/eclipse/appengine/localserver/ui/CloudSdkServerLaunchTab.java:
```diff
@@ -0,0 +1,158 @@
+  }
+
+  /**
+   * Returns the string in the text widget, or <code>null</code> if empty.
```
{@code } is the recommended way to do this now.

akerekes commented:
For changing extension point naming https://github.com/GoogleCloudPlatform/gcloud-eclipse-tools/issues/79 has been filed.

akerekes commented:
@briandealwis @elharo I've addressed all the comments above (or filed an issue if it seemed a larger chunk of work and not a requirement for this feature), and did a complete reformatting of the code. Please take another round of review.

elharo commented on plugins/com.google.cloud.tools.eclipse.appengine.localserver.test/src/com/google/cloud/tools/eclipse/appengine/localserver/facet/CloudSdkFacetUninstallDelegateTest.java:
```diff
@@ -0,0 +1,31 @@
+@RunWith(MockitoJUnitRunner.class)
+public class CloudSdkFacetUninstallDelegateTest {
+
+  @Mock
```
@Mock can be on same line

elharo commented on plugins/com.google.cloud.tools.eclipse.appengine.localserver.test/src/com/google/cloud/tools/eclipse/appengine/localserver/runtime/CloudSdkRuntimeTest.java:
```diff
@@ -0,0 +1,104 @@
+    when(runtime.getName()).thenReturn("");
+    IStatus status = cloudSdkRuntime.validate();
+    assertThat(status.getSeverity(), is(IStatus.ERROR));
+    assertThat(status.getMessage(), is(org.eclipse.wst.server.core.internal.Messages.errorRuntimeName));
```
better to use a literal string if it avoids a dependency, unless we expect this to change

elharo commented on plugins/com.google.cloud.tools.eclipse.appengine.localserver/src/com/google/cloud/tools/eclipse/appengine/localserver/CloudSdkUtils.java:
```diff
@@ -0,0 +1,120 @@
+   *
+   * @param project an Eclipse project
+   * @return true if {@code project} has the Cloud SDK facet and false
+   *         otherwise.
```
nit: no period since not a complete sentence

elharo commented on plugins/com.google.cloud.tools.eclipse.appengine.localserver/src/com/google/cloud/tools/eclipse/appengine/localserver/deploy/CloudSdkDeployProjectHandler.java:
```diff
@@ -0,0 +1,171 @@
+      return null;
+    }
+
+    final IRuntime runtime;
```
try combining these try blocks so the variable declarations aren't split. Short try blocks are a common anti-pattern.

elharo commented on plugins/com.google.cloud.tools.eclipse.appengine.localserver/src/com/google/cloud/tools/eclipse/appengine/localserver/deploy/CloudSdkDeployProjectHandler.java:
```diff
@@ -0,0 +1,171 @@
+      }
+    };
+    job.schedule();
+    return null;
```
does this ever return anything except null? If not, could just put return null into a single finally block

akerekes replied:
@elharo it returns null from many places, not only from catch blocks and the last statement, so I think after combining the try blocks into a single block, a finally block that replaces only 2 lines, do so much for readability. I don't think it's worth the change. If we could find a better way to organize the method to avoid those repeated *1. get value, 2. give error if null* blocks, that would be better.

briandealwis replied:
Seems like we'll be using code like this elsewhere too.  We could provide an abstract Handler that performs a lot of these validations.  Or we could soup up GCloudCommandDelegate and provide a UI helper, something more along the lines of
```
GCloudCommandDelegate commandLine = new GCloudCommandDelegate(project);
// prompt if required, throw OperationCanceledExeption on cancel, or CoreException on error
GCloudCommandUI.ensureValidLogin(shell, commandLine);
// Uses GCloudCommandDelegate.validate() returns an IStatus;
// supports adding new validators for different operations
GCloudCommandUI.validate(commandLine);
```

elharo commented on plugins/com.google.cloud.tools.eclipse.appengine.localserver/src/com/google/cloud/tools/eclipse/appengine/localserver/deploy/CloudSdkDeployProjectHandler.java:
```diff
@@ -0,0 +1,171 @@
+
+    final IPath[] fileSystemPath = new IPath[1];
+    PlatformUI.getWorkbench().getDisplay().syncExec(new Runnable() {
+      public void run() {
```
Might want to use @Override on these

elharo commented on plugins/com.google.cloud.tools.eclipse.appengine.localserver/src/com/google/cloud/tools/eclipse/appengine/localserver/server/CloudSdkLaunchConfigurationDelegate.java:
```diff
@@ -0,0 +1,233 @@
+                                                                  debugPort);
+      Activator.logInfo("Executing: " + commands);
+
+      Process p = null;
```
one line

elharo commented on plugins/com.google.cloud.tools.eclipse.appengine.localserver/src/com/google/cloud/tools/eclipse/appengine/localserver/facet/CloudSdkFacetUninstallDelegate.java:
```diff
@@ -0,0 +1,42 @@
+@SuppressWarnings("restriction") // For the FacetedProject in the javadoc
+public class CloudSdkFacetUninstallDelegate implements IDelegate {
+  /**
+   * Facet removal will already be handled by
```
" will already be handled" --> is handled

elharo commented on plugins/com.google.cloud.tools.eclipse.appengine.localserver/src/com/google/cloud/tools/eclipse/appengine/localserver/server/CloudSdkLaunchConfigurationDelegate.java:
```diff
@@ -0,0 +1,233 @@
+    }
+    IPath sdkLocation = runtime.getLocation();
+
+    CloudSdkServerBehaviour sb = (CloudSdkServerBehaviour) server.loadAdapter(CloudSdkServerBehaviour.class,
```
serverBehavior since sb sounds like a StringBuilder or StringBuffer

elharo commented on plugins/com.google.cloud.tools.eclipse.appengine.localserver/src/com/google/cloud/tools/eclipse/appengine/localserver/server/CloudSdkLaunchConfigurationDelegate.java:
```diff
@@ -0,0 +1,233 @@
+
+    final CloudSdkServer cloudSdkServer = CloudSdkServer.getCloudSdkServer(server);
+    int debugPort = -1;
+    if (mode.equals(ILaunchManager.DEBUG_MODE)) {
```
some devs like to put constants on the left side of .equals to avoid NullPointerExceptions

elharo commented on plugins/com.google.cloud.tools.eclipse.appengine.localserver/src/com/google/cloud/tools/eclipse/appengine/localserver/server/CloudSdkLaunchConfigurationDelegate.java:
```diff
@@ -0,0 +1,233 @@
+                                   projectName);
+
+    // Set JVM debugger connection parameters
+    Map<String, String> connectionParameters = new HashMap<String, String>();
```
HashMap<>

elharo commented on plugins/com.google.cloud.tools.eclipse.appengine.localserver/src/com/google/cloud/tools/eclipse/appengine/localserver/server/CloudSdkServer.java:
```diff
@@ -0,0 +1,215 @@
+  @Override
+  public IModule[] getChildModules(IModule[] module) {
+    // This is the same logic as in the GaeServer. The GaeServer would be
+    // replaced by the
```
please combine short lines

elharo commented on plugins/com.google.cloud.tools.eclipse.appengine.localserver/src/com/google/cloud/tools/eclipse/appengine/localserver/server/CloudSdkServer.java:
```diff
@@ -0,0 +1,215 @@
+    return new IModule[] { module };
+  }
+
+  @SuppressWarnings("unchecked")
```
might be able to remove this and fix warning directly

akerekes replied:
@elharo unfortunately the method in the parent class does not declare the generic type parameter

elharo commented on plugins/com.google.cloud.tools.eclipse.appengine.localserver/src/com/google/cloud/tools/eclipse/appengine/localserver/server/CloudSdkServer.java:
```diff
@@ -0,0 +1,215 @@
+  }
+
+  /**
+   * @return the default port on which to start the admin server.
```
nit: no period

elharo commented on plugins/com.google.cloud.tools.eclipse.appengine.localserver/src/com/google/cloud/tools/eclipse/appengine/localserver/server/CloudSdkServer.java:
```diff
@@ -0,0 +1,215 @@
+  }
+
+  /**
+   * @return the default port on which to start the API server.
```
nit: no period since not complete sentence

elharo commented on plugins/com.google.cloud.tools.eclipse.appengine.localserver/src/com/google/cloud/tools/eclipse/appengine/localserver/server/CloudSdkServerBehaviour.java:
```diff
@@ -0,0 +1,195 @@
+
+  @Override
+  public IStatus canStart(String launchMode) {
+    // Check that the port is not in use before can start
```
delete "can"

elharo commented on plugins/com.google.cloud.tools.eclipse.appengine.localserver/src/com/google/cloud/tools/eclipse/appengine/localserver/server/CloudSdkServerBehaviour.java:
```diff
@@ -0,0 +1,195 @@
+            try {
+              aLaunch.terminate();
+              manager.removeLaunch(aLaunch);
+              break;
```
why break here?

briandealwis replied:
Looks like it's iterating over the launches, so there's a possibility of a `ConcurrentModificationException`.

elharo commented on plugins/com.google.cloud.tools.eclipse.appengine.localserver/src/com/google/cloud/tools/eclipse/appengine/localserver/server/CloudSdkServerBehaviour.java:
```diff
@@ -0,0 +1,195 @@
+      @Override
+      public void handleDebugEvents(DebugEvent[] events) {
+        if (events != null) {
+          int size = events.length;
```
inline this

elharo commented on plugins/com.google.cloud.tools.eclipse.appengine.localserver/src/com/google/cloud/tools/eclipse/appengine/localserver/server/CloudSdkServerBehaviour.java:
```diff
@@ -0,0 +1,195 @@
+      new URL("http", server.getHostName(), server.getAdminPort(), "/quit").getContent();
+      try {
+        // TODO: confirm appropriate delay time
+        Thread.sleep(4000);
```
why do we delay here at all?

akerekes replied:
I'll look into it.

elharo commented on plugins/com.google.cloud.tools.eclipse.appengine.localserver/src/com/google/cloud/tools/eclipse/appengine/localserver/server/CloudSdkServerFlags.java:
```diff
@@ -0,0 +1,47 @@
+public class CloudSdkServerFlags {
+  private static final String FLAGS_FILE = "server-flags.json";
+
+  private static List<Flag> sFlags = null;
```
sFlags? I'm not sure what this is. Server flags? Please rename.

elharo commented on plugins/com.google.cloud.tools.eclipse.appengine.localserver/src/com/google/cloud/tools/eclipse/appengine/localserver/server/CloudSdkServerFlags.java:
```diff
@@ -0,0 +1,47 @@
+
+  private static List<Flag> sFlags = null;
+
+  public static List<Flag> getFlags() throws FileNotFoundException {
```
Please add JavaDoc indicating this is unmodifiable.

elharo commented on plugins/com.google.cloud.tools.eclipse.appengine.localserver/src/com/google/cloud/tools/eclipse/appengine/localserver/server/CloudSdkServerStartupChecker.java:
```diff
@@ -0,0 +1,120 @@
+import org.eclipse.wst.server.core.model.ServerBehaviourDelegate;
+
+/**
+ * Thread used to ping server to test when it is started.
```
This class is not a thread. Class that launches a thread to?

elharo commented on plugins/com.google.cloud.tools.eclipse.appengine.localserver/src/com/google/cloud/tools/eclipse/appengine/localserver/server/CloudSdkServerStartupChecker.java:
```diff
@@ -0,0 +1,120 @@
+    int count = 0;
+    try {
+      Thread.sleep(PING_DELAY);
+    } catch (Throwable e) {
```
per effective java, don't catch throwable

elharo commented on plugins/com.google.cloud.tools.eclipse.appengine.localserver/src/com/google/cloud/tools/eclipse/appengine/localserver/server/ServerFlagsInfo.java:
```diff
@@ -0,0 +1,52 @@
+
+  private List<Flag> flags = new ArrayList<Flag>();
+
+  public List<Flag> getFlags() {
```
danger; exposes mutable private state

elharo commented on plugins/com.google.cloud.tools.eclipse.appengine.localserver/src/com/google/cloud/tools/eclipse/appengine/localserver/server/server-flags.json:
```diff
@@ -0,0 +1,149 @@
+      {
+         "name":"datastore-consistency-policy",
+         "supportsArgument":"true",
+         "description":"The policy to apply when deciding whether a datastore write should appear in global queries"
```
"The policy to apply when deciding whether" --> "Whether"

elharo commented on plugins/com.google.cloud.tools.eclipse.appengine.localserver/src/com/google/cloud/tools/eclipse/appengine/localserver/server/server-flags.json:
```diff
@@ -0,0 +1,149 @@
+      {
+         "name":"datastore-path",
+         "supportsArgument":"true",
+         "description":"Path to a file used to store datastore contents (defaults to a file in --storage-path if not set)"
```
a file used to store --> that stores

elharo commented on plugins/com.google.cloud.tools.eclipse.appengine.localserver/src/com/google/cloud/tools/eclipse/appengine/localserver/server/server-flags.json:
```diff
@@ -0,0 +1,149 @@
+      {
+         "name":"logs-path",
+         "supportsArgument":"true",
+         "description":"Path to a file used to store request logs (defaults to a file in --storage-path if not set)"
```
used to store --> that stores

elharo commented on plugins/com.google.cloud.tools.eclipse.appengine.localserver/src/com/google/cloud/tools/eclipse/appengine/localserver/ui/CloudSdkServerLaunchTab.java:
```diff
@@ -0,0 +1,157 @@
+        case SWT.TRAVERSE_RETURN:
+        case SWT.TRAVERSE_TAB_NEXT:
+        case SWT.TRAVERSE_TAB_PREVIOUS:
+          if ((programFlagText.getStyle() & SWT.SINGLE) != 0) {
```
This sort of bit flag went out of style 20 years ago. Is there no other way to do this?

akerekes replied:
Apparently this is still the way to do it.

briandealwis replied:
If just interested in "enter", you can use the default select event (`SWT.DefaultSelection` or `SelectionListener#widgetDefaultSelected()`) on `Text` fields.

akerekes replied:
I have not yet touched/tested this code. First we should see what's the user experience and then see if that's sufficient, issue #83 has been filed to cover it.

elharo commented on plugins/com.google.cloud.tools.eclipse.appengine.localserver/src/com/google/cloud/tools/eclipse/appengine/localserver/ui/CloudSdkServerLaunchTab.java:
```diff
@@ -0,0 +1,157 @@
+          if ((programFlagText.getStyle() & SWT.SINGLE) != 0) {
+            e.doit = true;
+          } else {
+            if (!programFlagText.isEnabled() || (e.stateMask & SWT.MODIFIER_MASK) != 0) {
```
ditto

akerekes replied:
Apparently this is still the way to do it.

elharo commented on plugins/com.google.cloud.tools.eclipse.appengine.localserver/src/com/google/cloud/tools/eclipse/appengine/localserver/ui/ServerFlagSelectionDialog.java:
```diff
@@ -0,0 +1,151 @@
+    setMultipleSelection(false);
+
+    List<Flag> flags = null;
+    try {
```
move declaration and setElements call into try block

elharo commented on plugins/com.google.cloud.tools.eclipse.appengine.localserver/src/com/google/cloud/tools/eclipse/appengine/localserver/ui/ServerFlagSelectionDialog.java:
```diff
@@ -0,0 +1,151 @@
+    descriptionText = new Text(container, SWT.BORDER | SWT.WRAP | SWT.V_SCROLL);
+    descriptionText.setFont(container.getFont());
+    descriptionText.setEditable(false);
+    GridData gd = new GridData(GridData.FILL_HORIZONTAL);
```
gd --> gridData (Google style is not to abbreviate)

elharo commented on plugins/com.google.cloud.tools.eclipse.appengine.localserver/src/com/google/cloud/tools/eclipse/util/AdapterUtilities.java:
```diff
@@ -0,0 +1,49 @@
+    }
+
+    IAdapterManager adapterManager = Platform.getAdapterManager();
+    assert(adapterManager != null);
```
remove assert:

elharo commented on plugins/com.google.cloud.tools.eclipse.appengine.localserver/src/com/google/cloud/tools/eclipse/util/CustomMessageConsole.java:
```diff
@@ -0,0 +1,122 @@
+   * references to some of its active UI elements.
+   */
+  public static class ConsolePageParticipant implements IConsolePageParticipant {
+    public void activated() {
```
@Override?

elharo commented on plugins/com.google.cloud.tools.eclipse.appengine.localserver/src/com/google/cloud/tools/eclipse/util/CustomMessageConsole.java:
```diff
@@ -0,0 +1,122 @@
+       * may be waiting for the lock in setActiveToolbarManager.)
+       */
+      Display.getDefault().asyncExec(new Runnable() {
+        public void run() {
```
@Override

elharo commented on plugins/com.google.cloud.tools.eclipse.appengine.localserver/src/com/google/cloud/tools/eclipse/util/MavenUtils.java:
```diff
@@ -0,0 +1,38 @@
+      if (NatureUtils.hasNature(project, MavenUtils.MAVEN2_NATURE_ID)) {
+        return true;
+      }
+      if (NatureUtils.hasNature(project, MavenUtils.OLD_MAVEN2_NATURE_ID)) {
```
I agree that we probably want to get rid of the old nature.

elharo commented on plugins/com.google.cloud.tools.eclipse.appengine.localserver/src/com/google/cloud/tools/eclipse/util/NatureUtils.java:
```diff
@@ -0,0 +1,22 @@
+   * Returns {@code true} if the project is accessible and has the
+   * specified nature ID.
+   * 
+   * @param project
```
remove @param and @throws without descriptions

elharo commented on plugins/com.google.cloud.tools.eclipse.appengine.localserver/src/com/google/cloud/tools/eclipse/util/ProcessUtilities.java:
```diff
@@ -0,0 +1,266 @@
+  public static int launchProcessAndWaitFor(List<String> commands,
+                                            File workingDir,
+                                            final OutputStream outputStream,
+                                            IProcessReceiver processReceiver) throws InterruptedException,
```
maybe move throws clause to a single line

elharo commented on plugins/com.google.cloud.tools.eclipse.appengine.localserver/src/com/google/cloud/tools/eclipse/util/ProcessUtilities.java:
```diff
@@ -0,0 +1,266 @@
+                                            File workingDir,
+                                            final List<String> additionalPaths,
+                                            final OutputStream outputStream,
+                                            IProcessReceiver processReceiver) throws InterruptedException,
```
throws clause to a single line

elharo commented on plugins/com.google.cloud.tools.eclipse.appengine.localserver/src/com/google/cloud/tools/eclipse/util/ProcessUtilities.java:
```diff
@@ -0,0 +1,266 @@
+
+        public void run() {
+          BufferedReader processReader = 
+              new BufferedReader(new InputStreamReader(process.getInputStream()));
```
need to specify encoding for InputStreamReader. For that matter why do we use a reader here at all instead of just a stream?

elharo commented on plugins/com.google.cloud.tools.eclipse.appengine.localserver/src/com/google/cloud/tools/eclipse/util/ProcessUtilities.java:
```diff
@@ -0,0 +1,266 @@
+        public void run() {
+          BufferedReader processReader = 
+              new BufferedReader(new InputStreamReader(process.getInputStream()));
+          BufferedWriter outputWriter = new BufferedWriter(new OutputStreamWriter(outputStream));
```
speicfy encoding or just uise an outputstream and copy byte by byte

elharo commented on plugins/com.google.cloud.tools.eclipse.appengine.localserver/src/com/google/cloud/tools/eclipse/util/ProcessUtilities.java:
```diff
@@ -0,0 +1,266 @@
+  private static String copyInputStreamToString(InputStream inputStream) throws IOException {
+    ByteArrayOutputStream bos = new ByteArrayOutputStream();
+    ByteStreams.copy(inputStream, bos);
+    return new String(bos.toByteArray());
```
need to specify an encoding here

akerekes commented:
@elharo Comments regarding server-flags.json: this file contains an old version of the supported flags, it has not been touched as part of this PR. There's issue #74 to implement their usage and as part of that should this file be updated (or possibly get rid of if #70 is possible). Also as the description of the flags come from gcloud/dev_appserver.py it'd be better to have those values changed instead of us changing it on our end.

briandealwis commented on plugins/com.google.cloud.tools.eclipse.appengine.localserver/src/com/google/cloud/tools/eclipse/appengine/localserver/deploy/CloudSdkDeployProjectHandler.java:
```diff
@@ -0,0 +1,171 @@
+    }
+
+    final IPath[] fileSystemPath = new IPath[1];
+    PlatformUI.getWorkbench().getDisplay().syncExec(new Runnable() {
```
The activeShell should be passed in here, it's obtained from `HandlerUtil.getActiveShellChecked(event);`

elharo replied:
ping

akerekes replied:
The shell is obtained in the caller method and passed in via a parameter.

briandealwis replied:
`shell.getDisplay()`

elharo commented on plugins/com.google.cloud.tools.eclipse.appengine.localserver.test/src/com/google/cloud/tools/eclipse/appengine/localserver/facet/CloudSdkFacetUninstallDelegateTest.java:
```diff
@@ -0,0 +1,27 @@
+public class CloudSdkFacetUninstallDelegateTest {
+
+  @Mock private IProject project;
+  @Mock private IProjectFacetVersion fv;
```
facetVersion

elharo commented on plugins/com.google.cloud.tools.eclipse.appengine.localserver.test/src/com/google/cloud/tools/eclipse/appengine/localserver/runtime/CloudSdkRuntimeTest.java:
```diff
@@ -0,0 +1,104 @@
+@RunWith(MockitoJUnitRunner.class)
+public class CloudSdkRuntimeTest {
+  
+  @Mock
```
same line

elharo commented on plugins/com.google.cloud.tools.eclipse.appengine.localserver.test/src/com/google/cloud/tools/eclipse/appengine/localserver/server/CloudSdkServerTest.java:
```diff
@@ -0,0 +1,39 @@
+@RunWith(MockitoJUnitRunner.class)
+public class CloudSdkServerTest {
+
+  @Mock
```
same line

elharo commented on plugins/com.google.cloud.tools.eclipse.appengine.localserver.test/src/com/google/cloud/tools/eclipse/appengine/localserver/server/CloudSdkServerTest.java:
```diff
@@ -0,0 +1,39 @@
+  public void testGetCloudSdkServer_needsToLoadAdapter() {
+    when(server.getAdapter(CloudSdkServer.class)).thenReturn(null);
+    when(server.loadAdapter(eq(CloudSdkServer.class), any(IProgressMonitor.class)))
+      .thenReturn(mockCloudSdkServer);
```
indent + 2

elharo commented on plugins/com.google.cloud.tools.eclipse.appengine.localserver.test/src/com/google/cloud/tools/eclipse/appengine/localserver/util/ProcessUtilitiesTest.java:
```diff
@@ -0,0 +1,29 @@
+public class ProcessUtilitiesTest {
+
+	@Test
+	public void testlaunchProcessAndWaitFor_additionalPathAdded() throws InterruptedException, IOException {
```
indent - 4

elharo commented on plugins/com.google.cloud.tools.eclipse.appengine.localserver.test/src/com/google/cloud/tools/eclipse/appengine/localserver/util/ProcessUtilitiesTest.java:
```diff
@@ -0,0 +1,29 @@
+	@Test
+	public void testlaunchProcessAndWaitFor_additionalPathAdded() throws InterruptedException, IOException {
+		List<String> commands = Lists.newArrayList("/bin/bash", "-c", "echo $PATH");
+		File workingDir = new File("/tmp");
```
indent - 8

elharo commented on plugins/com.google.cloud.tools.eclipse.appengine.localserver.test/src/com/google/cloud/tools/eclipse/appengine/localserver/util/ProcessUtilitiesTest.java:
```diff
@@ -0,0 +1,29 @@
+		List<String> additionalPaths = Lists.newArrayList("/tmp/foo", "/tmp/bar");
+		ByteArrayOutputStream outputStream = new ByteArrayOutputStream();
+		ProcessUtilities.launchProcessAndWaitFor(commands, workingDir, additionalPaths, outputStream);
+		String result = outputStream.toString("utf-8");
```
UTF-8

elharo commented on plugins/com.google.cloud.tools.eclipse.appengine.localserver/META-INF/MANIFEST.MF:
```diff
@@ -15,5 +18,17 @@ Require-Bundle: org.eclipse.wst.server.core;bundle-version="1.7.0",
  org.eclipse.jst.common.project.facet.ui;bundle-version="1.4.510",
- javax.servlet;bundle-version="3.1.0";visibility:=reexport,
- javax.servlet.jsp;bundle-version="2.2.0";visibility:=reexport
+ javax.servlet;bundle-version="3.1.0",
```
2.5

elharo commented on plugins/com.google.cloud.tools.eclipse.appengine.localserver/src/com/google/cloud/tools/eclipse/appengine/localserver/GCloudCommandDelegate.java:
```diff
@@ -0,0 +1,178 @@
+   * @return true if the Cloud SDK and App Engine component have been installed
+   *         and false otherwise
+   */
+  public static boolean areCloudSdkAndAppEngineInstalled(String sdkLocation) throws IOException,
```
break before throws

elharo commented on plugins/com.google.cloud.tools.eclipse.appengine.localserver/src/com/google/cloud/tools/eclipse/appengine/localserver/server/CloudSdkServer.java:
```diff
@@ -0,0 +1,214 @@
+
+  @Override
+  public IModule[] getChildModules(IModule[] module) {
+    // This is the same logic as in the GaeServer. The GaeServer would be
```
does this comment need to be updated?

elharo commented on plugins/com.google.cloud.tools.eclipse.appengine.localserver/src/com/google/cloud/tools/eclipse/appengine/localserver/server/CloudSdkServerBehaviour.java:
```diff
@@ -0,0 +1,195 @@
+    CloudSdkServer server = CloudSdkServer.getCloudSdkServer(getServer());
+    int port = server.getApiPort();
+    if (!isPortAvailable(port)) {
+      String message = server.getServer().getName() + " has admin port set to "
```
looks like API port, not admin port

elharo commented on plugins/com.google.cloud.tools.eclipse.appengine.localserver/src/com/google/cloud/tools/eclipse/appengine/localserver/ui/CloudSdkRuntimeWizardFragment.java:
```diff
@@ -0,0 +1,203 @@
+        final IStatus runtimeStatus;
+        try {
+          runtimeStatus = event.getResult();
+        } catch (Exception e) {
```
catch a more specific Exception

elharo commented on plugins/com.google.cloud.tools.eclipse.appengine.localserver/src/com/google/cloud/tools/eclipse/appengine/localserver/ui/CloudSdkServerLaunchTab.java:
```diff
@@ -0,0 +1,159 @@
+import com.google.cloud.tools.eclipse.appengine.localserver.server.CloudSdkServer;
+
+/**
+ * A launch configuration tab that displays and edits "gcloud app preview"
```
gcloud preview app

elharo commented on plugins/com.google.cloud.tools.eclipse.appengine.localserver/src/com/google/cloud/tools/eclipse/appengine/localserver/ui/ServerFlagSelectionDialog.java:
```diff
@@ -0,0 +1,144 @@
+    setMessage("&Choose a variable (? = any character, * = any string):");
+    setMultipleSelection(false);
+
+    List<Flag> flags = null;
```
one line

elharo commented on plugins/com.google.cloud.tools.eclipse.appengine.localserver/src/com/google/cloud/tools/eclipse/util/CustomMessageConsole.java:
```diff
@@ -0,0 +1,128 @@
+   */
+  public static class ConsolePageParticipant implements IConsolePageParticipant {
+    @Override
+	public void activated() {
```
indent - 2

elharo commented on plugins/com.google.cloud.tools.eclipse.appengine.localserver/src/com/google/cloud/tools/eclipse/util/CustomMessageConsole.java:
```diff
@@ -0,0 +1,128 @@
+    }
+
+    @Override
+	public void deactivated() {
```
indent - 2

elharo commented on plugins/com.google.cloud.tools.eclipse.appengine.localserver/src/com/google/cloud/tools/eclipse/util/CustomMessageConsole.java:
```diff
@@ -0,0 +1,128 @@
+    }
+
+    @Override
+	public void init(IPageBookViewPage page, IConsole console) {
```
indent - 2

elharo commented on plugins/com.google.cloud.tools.eclipse.appengine.localserver/src/com/google/cloud/tools/eclipse/util/CustomMessageConsole.java:
```diff
@@ -0,0 +1,128 @@
+    }
+
+    @Override
+	@SuppressWarnings("unchecked")
```
indent - 2

elharo commented on plugins/com.google.cloud.tools.eclipse.appengine.localserver/src/com/google/cloud/tools/eclipse/util/CustomMessageConsole.java:
```diff
@@ -0,0 +1,128 @@
+       */
+      Display.getDefault().asyncExec(new Runnable() {
+        @Override
+		public void run() {
```
indent - 8

elharo commented on plugins/com.google.cloud.tools.eclipse.appengine.localserver/src/com/google/cloud/tools/eclipse/util/NatureUtils.java:
```diff
@@ -0,0 +1,19 @@
+   * 
+   * @return {@code true} if the project is accessible and has the
+   *         specified nature ID
+   * @throws CoreException
```
no @throws without a message

elharo commented on plugins/com.google.cloud.tools.eclipse.appengine.localserver/src/com/google/cloud/tools/eclipse/util/ProcessUtilities.java:
```diff
@@ -0,0 +1,236 @@
+   * @param commands commands to pass to the {@link ProcessBuilder}
+   * @param workingDir directory to use as the working directory
+   * @param additionalPaths list of additional directories to be appended to the
+   *          PATH environmental variable
```
environment variable

elharo commented on plugins/com.google.cloud.tools.eclipse.appengine.localserver/src/com/google/cloud/tools/eclipse/util/ProcessUtilities.java:
```diff
@@ -0,0 +1,236 @@
+    moveClasspathArgToEnvironmentVariable(commands, pb);
+
+    // if given a non-null, non-empty list of paths, then append to the PATH
+    // environmental variable
```
environment

elharo commented on plugins/com.google.cloud.tools.eclipse.appengine.localserver/src/com/google/cloud/tools/eclipse/util/ProcessUtilities.java:
```diff
@@ -0,0 +1,236 @@
+          try {
+            int bytesRead = 0;
+            byte[] buf = new byte[1024];
+			while ((bytesRead = inputStream.read(buf)) != -1) {
```
deindent

elharo commented on plugins/com.google.cloud.tools.eclipse.appengine.localserver/src/com/google/cloud/tools/eclipse/util/ProcessUtilities.java:
```diff
@@ -0,0 +1,236 @@
+  private static String copyInputStreamToString(InputStream inputStream) throws IOException {
+    ByteArrayOutputStream bos = new ByteArrayOutputStream();
+    ByteStreams.copy(inputStream, bos);
+    return new String(bos.toByteArray(), Charset.forName("utf-8"));
```
UTF-8

briandealwis replied:
Better yet, `StandardCharsets.UTF_8`.

akerekes commented:
@briandealwis @elharo PTAL

briandealwis commented on plugins/com.google.cloud.tools.eclipse.appengine.localserver.test/src/com/google/cloud/tools/eclipse/appengine/localserver/util/ProcessUtilitiesTest.java:
```diff
@@ -0,0 +1,29 @@
+public class ProcessUtilitiesTest {
+
+  @Test
+  public void testlaunchProcessAndWaitFor_additionalPathAdded() throws InterruptedException, IOException {
```
This test won't work on Win32.

akerekes replied:
#87

briandealwis commented on plugins/com.google.cloud.tools.eclipse.appengine.localserver/plugin.properties:
```diff
@@ -0,0 +1,14 @@
+runtimeTypeVendor=Google
+
+serverTypeName=Google Cloud SDK
+serverTypeDescription=App Engine DevAppServer2
```
I don't think many people know what `DevAppServer2` refers to

briandealwis commented on plugins/com.google.cloud.tools.eclipse.appengine.localserver/plugin.xml:
```diff
@@ -100,4 +100,205 @@
+         point="org.eclipse.debug.ui.launchConfigurationTypeImages">
+      <launchConfigurationTypeImage
+            configTypeID="com.google.cloud.tools.eclipse.CloudSdkLaunchConfigurationType"
+            icon="icons/gcp.png"
```
Not the AppEngine icon?

akerekes replied:
#84

briandealwis commented on plugins/com.google.cloud.tools.eclipse.appengine.localserver/plugin.xml:
```diff
@@ -100,4 +100,205 @@
+   <extension
+         point="org.eclipse.wst.server.ui.serverImages">
+      <image
+            icon="icons/gcp.png"
```
Not the AppEngine icon?

akerekes replied:
#84

briandealwis commented on plugins/com.google.cloud.tools.eclipse.appengine.localserver/plugin.xml:
```diff
@@ -100,4 +100,205 @@
+                        value="1">
+                  </count>
+                  <iterate
+                        operator="and">
```
You could test for `adapt` to `IProject` to short-circuit the property tester.

akerekes replied:
The deploy code does not work, see: #66 
I've added this comment to the issue.

briandealwis commented on plugins/com.google.cloud.tools.eclipse.appengine.localserver/src/com/google/cloud/tools/eclipse/appengine/localserver/Activator.java:
```diff
@@ -0,0 +1,124 @@
+   * @param title the dialog's title
+   * @param message the message
+   */
+  public static void logAndDisplayError(final Shell parent,
```
The better approach is to use the JFace `org.eclipse.ui.statushandler.StatusHandlers` class

akerekes replied:
#85

briandealwis commented on plugins/com.google.cloud.tools.eclipse.appengine.localserver/src/com/google/cloud/tools/eclipse/appengine/localserver/GCloudCommandDelegate.java:
```diff
@@ -0,0 +1,176 @@
+                                           String mode,
+                                           String apiHost,
+                                           int apiPort,
+                                           int debugPort) throws NullPointerException, InvalidPathException {
```
NPE isn't declared further up

briandealwis commented on plugins/com.google.cloud.tools.eclipse.appengine.localserver/src/com/google/cloud/tools/eclipse/appengine/localserver/facet/CloudSdkFacetInstallDelegate.java:
```diff
@@ -0,0 +1,53 @@
+   */
+  @Override
+  public void execute(IProject project,
+                      IProjectFacetVersion fv,
```
`fv` &rarr; `facetVersion`

briandealwis commented on plugins/com.google.cloud.tools.eclipse.appengine.localserver/src/com/google/cloud/tools/eclipse/appengine/localserver/facet/CloudSdkFacetUninstallDelegate.java:
```diff
@@ -0,0 +1,42 @@
+   */
+  @Override
+  public void execute(IProject project,
+                      IProjectFacetVersion fv,
```
`fv` &rarr; `facetVersion`

briandealwis commented on plugins/com.google.cloud.tools.eclipse.appengine.localserver/src/com/google/cloud/tools/eclipse/appengine/localserver/server/CloudSdkServer.java:
```diff
@@ -0,0 +1,213 @@
+    if (module[0] != null && module[0].getModuleType() != null) {
+      IModule thisModule = module[module.length - 1];
+      IModuleType moduleType = thisModule.getModuleType();
+      if (moduleType != null && "jst.ear".equals(moduleType.getId())) { //$NON-NLS-1$
```
`"jst.ear"` and `"jst.web"` should be constants

elharo replied:
Personal preference but in general, I don't see a lot of value to turning constant strings into named constants. "jst.ear" is as descriptive as JST_EAR, maybe a bit more so. Numeric constants are a different story, of course.

briandealwis replied:
Keep misspellings in one place ;)  More seriously, one place to check and it makes validations like your plugin.xml unit tests much more meaningful.

We can also give these more meaningful names. "jst.web" doesn't scream out "Servlet API" to me.

briandealwis commented on plugins/com.google.cloud.tools.eclipse.appengine.localserver/src/com/google/cloud/tools/eclipse/appengine/localserver/server/CloudSdkServerFlags.java:
```diff
@@ -0,0 +1,47 @@
+import com.google.gson.GsonBuilder;
+
+/**
+ * Holds the list of "gcloud app run" flags for the Cloud SDK server.
```
Since we're executing `gcloud` for a bunch of other reasons can't we do that here too?

akerekes replied:
We're executing dev_appserver.py (so the javadoc is incorrect), but that's beside the point. The output of the command contains several flags used by other runtimes (python, php, go?), and also not very machine processing friendly (like json). Nevertheless, there's an issue for this, might need some changes to the description: #70

briandealwis commented on plugins/com.google.cloud.tools.eclipse.appengine.localserver/src/com/google/cloud/tools/eclipse/appengine/localserver/server/CloudSdkServerStartupChecker.java:
```diff
@@ -0,0 +1,121 @@
+   * Starts a thread that pings the url to be monitored.
+   */
+  public void start() {
+    Thread t = new Thread(NAME) {
```
This should be a Job. They can be rescheduled.

akerekes replied:
#86

briandealwis commented on plugins/com.google.cloud.tools.eclipse.appengine.localserver/src/com/google/cloud/tools/eclipse/util/CustomMessageConsole.java:
```diff
@@ -0,0 +1,128 @@
+ * A message console with additional features, such as displaying a terminate
+ * button.
+ */
+public class CustomMessageConsole extends MessageConsole {
```
Do we need this having figured out the process factory connection?

akerekes replied:
It is referred by the deploy code (that is broken right now, see #66 )
I'd rather keep it around until the deploy is implemented properly and getting rid of it then if it is not needed, but open to delete it and all deploy related code. WDYT @briandealwis @elharo ?

akerekes commented:
@briandealwis @elharo PTAL

elharo commented on plugins/com.google.cloud.tools.eclipse.appengine.localserver/plugin.properties:
```diff
@@ -0,0 +1,14 @@
+runtimeTypeName=GAE Standard
+runtimeTypeVendor=Google
+
+serverTypeName=GAE Standard Local Server
```
App Engine Standard

elharo commented on plugins/com.google.cloud.tools.eclipse.appengine.localserver/src/com/google/cloud/tools/eclipse/appengine/localserver/ui/CloudSdkRuntimeWizardFragment.java:
```diff
@@ -0,0 +1,196 @@
+    button.addSelectionListener(new SelectionAdapter() {
+      @Override
+	  public void widgetSelected(SelectionEvent event) {
+        DirectoryDialog dlg = new DirectoryDialog(composite.getShell(), SWT.OPEN);
```
dialog or directoryDialog

elharo commented on plugins/com.google.cloud.tools.eclipse.appengine.localserver/src/com/google/cloud/tools/eclipse/appengine/localserver/ui/CloudSdkRuntimeWizardFragment.java:
```diff
@@ -0,0 +1,196 @@
+  }
+
+  private CloudSdkRuntime getRuntimeDelegate() {
+    IRuntimeWorkingCopy wc = (IRuntimeWorkingCopy) getTaskModel().getObject(TaskModel.TASK_RUNTIME);
```
workingCopy

elharo commented:
LGTM from me. I think we're at the point where we need to commit this and iterate on it to make further progress.

briandealwis commented:
Agreed. LGTM.

> On 21-Apr-2016, at 12:58 PM, Elliotte Rusty Harold <notifications@github.com> wrote:
> 
> LGTM from me. I think we're at the point where we need to commit this and iterate on it to make further progress.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub <https://github.com/GoogleCloudPlatform/gcloud-eclipse-tools/pull/65#issuecomment-213015729>
